### PR TITLE
Add OpenTelemetry tracing

### PR DESCRIPTION
**Pull request by andmat900**
### Applicable Issues

- https://github.com/eiffel-community/etos/issues/228
- Depends On https://github.com/eiffel-community/etos-library/pull/27

### Description of the Change

This change adds OpenTelemetry instrumentation to etos-environment-provider.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by: Andrei Matveyeu, andrei.matveyeu@axis.com

### CODE REVIEW

File: src/environment_provider/environment_provider.py [hunk 1]
```diff
@@ -78,6 +80,8 @@ def __init__(self, suite_id: str, suite_runner_ids: list[str], copy: bool = True
         """
         FORMAT_CONFIG.identifier = suite_id
         self.logger.info("Initializing EnvironmentProvider task.")
+        self.tracer = opentelemetry.trace.get_tracer(__name__)
+        self.tracer_context = get_current_context()
```
Comment (fredjn) [inline src/environment_provider/environment_provider.py:84]: I can't seem to find any usage of the `tracer_context`-variable, but I might have missed something?

Comment (fredjn) [reply src/environment_provider/environment_provider.py:84]: This is also the only usage of `get_current_context` ?

File: src/iut_provider/utilities/external_provider.py [hunk 2]
```diff
@@ -108,6 +110,13 @@ def checkin(self, iut: Iut) -> None:
         iuts = [iut.as_dict for iut in iut]
 
         host = self.ruleset.get("stop", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
+        if otel_span is not None:
```
Comment (fredjn) [inline src/iut_provider/utilities/external_provider.py:115]: Same as previous comments

File: src/log_area_provider/utilities/external_provider.py [hunk 3]
```diff
@@ -108,6 +110,13 @@ def checkin(self, log_area: LogArea) -> None:
         log_areas = [log_area.as_dict for log_area in log_area]
 
         host = self.ruleset.get("stop", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
```
Comment (fredjn) [inline src/log_area_provider/utilities/external_provider.py:114]: Same as previous comments

File: src/log_area_provider/utilities/external_provider.py [hunk 4]
```diff
@@ -158,11 +167,19 @@ def start(self, minimum_amount: int, maximum_amount: int) -> str:
             "dataset": self.dataset.get("dataset"),
             "context": self.dataset.get("context"),
         }
+        host = self.ruleset.get("start", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
+        if otel_span is not None:
```
Comment (fredjn) [inline src/log_area_provider/utilities/external_provider.py:173]: Same as previous comments

File: src/environment_provider/environment.py [hunk 5]
```diff
@@ -75,21 +79,30 @@ def release_environment(
...
+        if not success:
+            failure = exception
+
+    span_name = "stop_log_area"
+    with TRACER.start_as_current_span(span_name):
+        success, exception = checkin_provider(
+            LogArea(**log_area), LogAreaProvider(etos, jsontas, log_area_ruleset)
+        )
+        if not success:
+            failure = exception
```
Comment (t-persson) [inline src/environment_provider/environment.py:95]: We should record this exception:
https://opentelemetry-python.readthedocs.io/en/latest/api/trace.span.html#opentelemetry.trace.span.Span.record_exception

File: src/environment_provider/environment.py [hunk 6]
```diff
@@ -75,21 +79,30 @@ def release_environment(
...
+
+    span_name = "stop_execution_space"
+    with TRACER.start_as_current_span(span_name):
+        success, exception = checkin_provider(
+            ExecutionSpace(**executor),
+            ExecutionSpaceProvider(etos, jsontas, executor_ruleset),
+        )
+        if not success:
+            failure = exception
```
Comment (t-persson) [inline src/environment_provider/environment.py:104]: We should record this exception:
https://opentelemetry-python.readthedocs.io/en/latest/api/trace.span.html#opentelemetry.trace.span.Span.record_exception

File: src/environment_provider/environment.py [hunk 7]
```diff
@@ -75,21 +79,30 @@ def release_environment(
...
-        ExecutionSpace(**executor),
-        ExecutionSpaceProvider(etos, jsontas, executor_ruleset),
-    )
-    if not success:
-        failure = exception
+    span_name = "stop_iuts"
+    with TRACER.start_as_current_span(span_name):
+        success, exception = checkin_provider(Iut(**iut), IutProvider(etos, jsontas, iut_ruleset))
+        if not success:
+            failure = exception
```
Comment (t-persson) [inline src/environment_provider/environment.py:87]: We should record this exception:
https://opentelemetry-python.readthedocs.io/en/latest/api/trace.span.html#opentelemetry.trace.span.Span.record_exception

File: src/execution_space_provider/utilities/external_provider.py [hunk 8]
```diff
@@ -218,11 +231,13 @@ def wait(self, provider_id: str) -> dict:
                 first_iteration = False
             else:
                 time.sleep(2)
+            headers = {"X-ETOS-ID": self.identifier}
+            self.logger.info("OpenTelemetry context headers: %s", headers)
```
Comment (t-persson) [inline src/execution_space_provider/utilities/external_provider.py:235]: We don't use opentelemetry here

File: src/execution_space_provider/utilities/external_provider.py [hunk 9]
```diff
@@ -186,11 +192,18 @@ def start(self, minimum_amount: int, maximum_amount: int) -> str:
             "dataset": self.dataset.get("dataset"),
             "context": self.dataset.get("context"),
         }
+        host = self.ruleset.get("start", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
+        otel_span.set_attribute("request.host", host)
+        otel_span.set_attribute("request.headers", json.dumps(headers, indent=4))
+        otel_span.set_attribute("request.body", json.dumps(data, indent=4))
+        self.logger.info("OpenTelemetry context headers: %s", headers)
```
Comment (t-persson) [inline src/execution_space_provider/utilities/external_provider.py:201]: Nit: Necessary?

File: src/execution_space_provider/utilities/external_provider.py [hunk 10]
```diff
@@ -114,6 +116,12 @@ def checkin(self, execution_space: ExecutionSpace) -> None:
         execution_spaces = [execution_space.as_dict for execution_space in execution_space]
 
         host = self.ruleset.get("stop", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
+        otel_span.set_attribute("request.host", host)
+        otel_span.set_attribute("request.headers", json.dumps(headers, indent=4))
+        otel_span.set_attribute("request.body", json.dumps(execution_spaces, indent=4))
+        self.logger.info("OpenTelemetry context headers: %s", headers)
```
Comment (t-persson) [inline src/execution_space_provider/utilities/external_provider.py:124]: Nit: Necessary?

File: src/execution_space_provider/utilities/external_provider.py [hunk 11]
```diff
@@ -114,6 +116,12 @@ def checkin(self, execution_space: ExecutionSpace) -> None:
         execution_spaces = [execution_space.as_dict for execution_space in execution_space]
 
         host = self.ruleset.get("stop", {}).get("host")
+        headers = {"X-ETOS-ID": self.identifier}
+        otel_span = opentelemetry.trace.get_current_span()
+        otel_span.set_attribute("request.host", host)
+        otel_span.set_attribute("request.headers", json.dumps(headers, indent=4))
+        otel_span.set_attribute("request.body", json.dumps(execution_spaces, indent=4))
```
Comment (t-persson) [inline src/execution_space_provider/utilities/external_provider.py:123]: Should we really indent the JSON here? It makes the request to the OTEL collector larger.
This comment applies to all indent=4 dumps

**Review by t-persson — CHANGES_REQUESTED** [review]

File: src/environment_provider/environment.py (hunk 5 above)
Comment (andmat900) [reply src/environment_provider/environment.py:99]: Done

File: src/environment_provider/environment.py (hunk 6 above)
Comment (andmat900) [reply src/environment_provider/environment.py:110]: Done

File: src/execution_space_provider/utilities/external_provider.py (hunk 11 above)
Comment (andmat900) [reply src/execution_space_provider/utilities/external_provider.py]: If this isn't done, the JSON is shown in Jaeger as a string without pretty formatting.

File: src/environment_provider/environment.py (hunk 7 above)
Comment (andmat900) [reply src/environment_provider/environment.py:89]: Done

File: src/execution_space_provider/utilities/external_provider.py (hunk 11 above)
Comment (fredjn) [reply src/execution_space_provider/utilities/external_provider.py]: I'm leaning towards keeping the amount of data down, and If we feel it's a problem to re-format the json (when/if we inspect it) we can re-evaluate and refactor.

**Review by fredjn — APPROVED** [review]

File: src/execution_space_provider/utilities/external_provider.py (hunk 11 above)
Comment (andmat900) [reply src/execution_space_provider/utilities/external_provider.py]: Pretty formatting is now removed.

File: src/environment_provider/environment_provider.py [hunk 12]
```diff
@@ -414,14 +417,17 @@ def checkout(
...
-            # Check out and assign IUTs to test runners.
-            iuts = self.iut_provider.wait_for_and_checkout_iuts(
-                minimum_amount=1,
-                maximum_amount=self.dataset.get(
-                    "maximum_amount", self.etos.config.get("TOTAL_TEST_COUNT")
-                ),
-            )
-            self.splitter.assign_iuts(test_runners, iuts)
+
+            with self.tracer.start_as_current_span("request_iuts") as span:
```
[the remark this answers is not in the document]
Comment (andmat900) [reply src/environment_provider/environment_provider.py]: The latest update includes mandatory attributes. Note that it increases the amount of instrumentation code. A better way might be to encapsulate instrumentation somewhere else, like etos_library.

File: src/log_area_provider/utilities/external_provider.py [hunk 13]
```diff
@@ -114,8 +123,8 @@ def checkin(self, log_area: LogArea) -> None:
         headers = {"X-ETOS-ID": self.identifier}
         TraceContextTextMapPropagator().inject(headers)
         otel_span = opentelemetry.trace.get_current_span()
-        otel_span.set_attribute("http.request.host", host)
-        otel_span.set_attribute("http.request.body", json.dumps(log_areas, indent=4))
+        otel_span.set_attribute(SpanAttributes.URL_FULL, host)
```
Comment (fredjn) [inline src/log_area_provider/utilities/external_provider.py:126]: Perhaps rename to span to keep naming consistent?

Comment (andmat900) [reply src/log_area_provider/utilities/external_provider.py]: Changed the name.

**Pull request merged**